### PR TITLE
Add prehooks to put and delete

### PR DESCRIPTION
**Pull request by dozyio**
* Add prehooks to `put` and `delete` so that actions can be triggered before the elements are added / removed

* Miscellaneous tidy up

### CODE REVIEW

File: crdt.go [hunk 1]
```diff
@@ -77,19 +77,30 @@ type SessionDAGService interface {
 type Options struct {
 	Logger              logging.StandardLogger
 	RebroadcastInterval time.Duration
-	// The PutHook function is triggered whenever an element
+	// The PutPreHook function is triggered before an element
+	// is added to the datastore (either by a local or remote update),
+	// and only when that addition is considered the prevalent value.
+	PutPreHook func(k ds.Key, v []byte) error
```
Comment (hsanjuan) [inline crdt.go:83]: Per the code below, the prehook is triggered in all cases, even when it is not going to be the prevalent value.

File: crdt.go [hunk 2]
```diff
@@ -77,19 +77,30 @@ type SessionDAGService interface {
...
-	PutHook func(k ds.Key, v []byte)
-	// The DeleteHook function is triggered whenever a version of an
+	PutPostHook func(k ds.Key, v []byte)
+	// The DeletePreHook function is triggered before an element is removed
+	// from the datastore (either by a local or remote update).
+	// Unordered and concurrent updates may result in the DeletePreHook being
+	// triggered even though the element is still present in the datastore
+	// because it was re-added. If that is relevant, use Has() to check if the
+	// removed element is still part of the datastore.
+	DeletePreHook func(k ds.Key) error
```
Comment (hsanjuan) [inline crdt.go:95]: Explanation here is weird. I would expect that when a delete happens and a pre-hook is triggered, the element is still in the datastore since it is a PRE-hook. I kind of get what it means, but Has() should generally return true because the tombstone that we are writing has not been written yet, so there will be at least one missing tombstone for the value, unless we are re-adding a tombstone.

**Review by hsanjuan — CHANGES_REQUESTED** [review]
Hello, thanks for the contribution. Would you be able to provide some rationale as to why these pre-hooks are useful to you?

I'm not convinced about calling them "hooks" as existing post-hooks are akin to notifications while these pre-hooks allow side-effects or aborting set operations when returning error.

I'm not sure this is properly wired either, `putElems` will abort writing elements when the setPreHook fails. Same for `putTombs`, so `Merge` is going to be returning errors and leave things half-written when a hook returns error for a single element.

This in turn will leave [merkle-crdt nodes unprocessed and in the to-be-processed set, so more attempts to merge may happen and if I'm not mistaken probably this happens indefinitely](https://github.com/ipfs/go-ds-crdt/blob/f8ca30f0e8565d7204078fe29452111b84a0d1a2/crdt.go#L783).

I understand that rather than hooks these are more like filters or validators for incoming elements/tombstones (i.e. not allowing writing invalid keys). If that is the case, while the node can maintain certain local validity with these in case of other peers sending invalid values, this does not prevent other peers from misbehaving and adding invalid stuff to the DAG, so that must at least be kept in mind.

In short:

- Explain rationale so that we can make sure it makes sense to take the extra code
- If so, wiring it in needs to happen more carefully, probably without bubbling errors, or bubbling in a different way (discriminating between error types so that an error is not interpreted as something needing to be retried).

**Comment by hsanjuan** [comment]
>  Half thinking trusted peers for writing / untrusted for reads but maybe that could be handled at the libp2p layer.

You could handle it in the pubsub validator... so that you ignore any pubsub message that doesn't comply to your requirements. Perhaps you need to make your pubsub validator aware of your message structure (i.e. the deltas), and maybe that means parsing things twice, once in the validator, and once during processing, but you could gate things at that stage. (that is where updates from untrusted peers are filtered out too).
